backbone/src/test/java/org/researchstack/backbone/ui/task/TaskViewModelTest.kt

```diff
@@ -0,0 +1,419 @@
+
+    @Rule
+    @JvmField
+    val rule: TestRule = InstantTaskExecutorRule()
+
+    @Mock
+    private var application: Application? = null
+    @Mock
+    private var intent: Intent? = null
+    @Mock
+    private var task: Task? = null
+
+    private lateinit var taskViewModel: TaskViewModel
+
+    @Before
+    fun before() {
+        whenever(intent?.getSerializableExtra(TaskActivity.EXTRA_TASK)).thenReturn(eq(task))
+    }
+
+    @Test
+    fun testInitViewModel_TaskShouldNotBeEmpty_EditModeMustBeFalse() {
+        // Assemble
+        // Act
+        taskViewModel = createViewModel()
+
+        // Assert
+        assertEquals(taskViewModel.editing, false)
+        assertEquals(taskViewModel.task, task)
+    }
+
+    @Test
+    fun testEdit_editShouldBeTrue_EditStepShouldBeUpdated_CancelMenuShouldBeTrueToHideCancelInTitleBar() {
+        // Assemble
+        val selectedEditStep = createCurrentStep("selectedEditStep")
+        taskViewModel = createViewModel()
+
+        // Act
+        taskViewModel.edit(selectedEditStep)
+
+        // Assert
+        Assert.assertEquals(taskViewModel.currentStep, selectedEditStep)
+        Assert.assertEquals(taskViewModel.editStep.value, selectedEditStep)
+        Assert.assertEquals(taskViewModel.editing, true)
+        Assert.assertEquals(taskViewModel.hideMenuItemCancel.value, true)
+    }
+
+
+    @Test
+    fun testRemoveUpdatedLayout_editShouldBeFalse_EditStepShouldBeUpdated_CancelMenuShouldBeFalseToShowCancelInTitleBar() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
+        val currentStepMocked = createCurrentStep()
+        val nextStepMocked = createCurrentStep("nextStepMocked")
+        taskViewModel.currentStep = currentStepMocked
+        `when`(task!!.getStepAfterStep(any(), any())).thenReturn(nextStepMocked)
+        `when`(taskViewModel.isReviewStep(nextStepMocked)).thenReturn(true)
+
+        // Act
+        taskViewModel.removeUpdatedLayout()
+
+        // Assert
+        Assert.assertEquals(taskViewModel.editing, false)
+        Assert.assertEquals(taskViewModel.updateCancelEditInLayout.value, true)
+        Assert.assertEquals(taskViewModel.hideMenuItemCancel.value, false)
+    }
+
+    @Test
+    fun testNext_ShouldUpdateCurrentStepToNextStep_shouldUpdateCurrentStepEventToNextStep() {
+        // Assemble
+        taskViewModel = createViewModel()
+        val currentStepMocked = createCurrentStep()
+        val nextStepMocked = createCurrentStep("nextStepMocked")
+        taskViewModel.currentStep = currentStepMocked
+        `when`(task!!.getStepAfterStep(any(), any())).thenReturn(nextStepMocked)
+
+        // Act
+        taskViewModel.nextStep()
+
+        // Assert
+        Assert.assertEquals(taskViewModel.currentStep, nextStepMocked)
+        Assert.assertEquals(taskViewModel.currentStepEvent.value!!.step, nextStepMocked)
+    }
+
+    @Test
+    fun testPrevious_ShouldUpdateCurrentStepToPreviousStep_shouldUpdateCurrentStepEventToPreviousStep() {
```

Comment:
Same rename here (and I won't add a comment for all tests) but when possible, prefer to use the full method name you're testing. It's not a hard rule, and in some simple tests and classes this is not needed, but this one is big, lots of tests. 

So `testPreviousStep_...`